File: 36两个链表的第一个公共结点.js
/*function ListNode(x){
    this.val = x;
    this.next = null;
}*/
function FindFirstCommonNode(pHead1, pHead2)
{
    // write code here
    let node1 = pHead1;
    while(node1) {
        node1.first = true;
        node1 = node1.next;
    }
    let node2 = pHead2;
    while(node2) {
        if(node2.first) {
            return node2;
        }
        node2 = node2.next;
    }
}
// javascript 不需要句尾加分号
// javascript 允许动态的向function里增加成员
// javascript function 如果没有返回任何值 则默认返回 null

File: 剑指offer全解/36两个链表的第一个公共结点.py
class Solution:
    def FindFirstCommonNode(self, pHead1, pHead2):
        # write code here
        p1 = pHead1
        p2 = pHead2
        size1 = size2 = 0
        while p1:
            size1 += 1
            p1 = p1.next
        while p2:
            size2 += 1
            p2 = p2.next
        if size1 < size2:
            pHead1, pHead2 = pHead2, pHead1
        distance = size1 - size2
        for i in range(distance):
            pHead1 = pHead1.next
        while pHead1 != pHead2:
            pHead1 = pHead1.next
            pHead2 = pHead2.next
        return pHead1